torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -418,19 +433,100 @@ void ONNXSetDynamicInputShape(
   }
 }
 
+bool SequenceTypeOutput(Node* node) {
+  if (node->kind() == ::c10::onnx::SplitToSequence ||
+      node->kind() == ::c10::onnx::SequenceInsert ||
+      node->kind() == ::c10::onnx::SequenceEmpty ||
+      node->kind() == ::c10::onnx::SequenceConstruct)
+    return true;
+  return false;
+}
+
+void ONNXInferOutputShapeType(
+    Value* graph_output,
+    at::Tensor output,
+    bool onnx_shape_inference) {
+  if (onnx_shape_inference) {
+    graph_output->setType(
+        MergeInferredType(TensorType::create(output), graph_output->type()));
+  } else {
+    graph_output->inferTypeFrom(output);
+  }
+}
+
 void ONNXAssignOutputShape(
     std::shared_ptr<Graph>& graph,
     at::ArrayRef<at::Tensor> outputs,
+    const python::IODescriptor& desc,
     bool onnx_shape_inference) {
-  TORCH_INTERNAL_ASSERT(graph->outputs().size() == outputs.size());
-  for (size_t i = 0; i < outputs.size(); ++i) {
-    if (onnx_shape_inference) {
-      graph->outputs()[i]->setType(MergeInferredType(
-          TensorType::create(outputs[i]), graph->outputs()[i]->type()));
+  auto vars_it = outputs.begin();
+  auto vars_it_end = outputs.end();
+  auto desc_it = desc.structure.begin();
+  auto str_it = desc.strings.begin();
+  auto str_end = desc.strings.end();
+
+  char type = *desc_it;
+  size_t graph_output_index = 0;
+  size_t graph_outputs_size = graph->outputs().size();
+  size_t outputs_index = 0;
+
+  while (outputs_index < outputs.size() &&
+         graph_output_index < graph_outputs_size) {
+    bool sequence_type_output =
+        SequenceTypeOutput(graph->outputs()[graph_output_index]->node());
+    if (sequence_type_output) {
+      // "Outputs" is a flattened list of tensors, but
+      //  graph outputs are not flattened. Unflatten
+      // "outputs" to assign shape for the corresponding item.
+      if (type == D::TupleOpen) {
+        desc_it++;
+        while (*desc_it++ != D::TupleClose) {
+          TORCH_INTERNAL_ASSERT(*desc_it == D::Variable);
+          ONNXInferOutputShapeType(
+              graph->outputs()[graph_output_index],
+              outputs[outputs_index],
```

Comment:
is `graph->outputs()[graph_output_index]` sequence? on the other hand `outputs[outputs_index]` tensor? how do these two work together

Reply:
Updated this part of the code.